src/verifier/verifier.rs

```diff
@@ -19,7 +27,535 @@ pub fn build_symbol_table(ast: &mut Whamm, err: &mut ErrorGen) -> SymbolTable {
+                        false,
+                        "Can't get type of expr of this unary operation".to_owned(),
+                        &<Option<crate::parser::types::Location> as Clone>::clone(expr.loc())
+                            .map(|l| l.line_col),
+                    );
+                    None
+                }
+            }
+            Expr::Call {
+                fn_target,
+                args,
+                loc,
+            } => {
+                // TODO: finish type checking for function calls
+                // lookup type of function
+                let mut param_tys = vec![];
+                if let Some(args) = args {
+                    for arg in args {
+                        if let Some(ty) = self.visit_expr(arg) {
+                            param_tys.push(ty);
+                        } else {
+                            self.err.type_check_error(
+                                false,
+                                "Can't get type of argument".to_owned(),
+                                &<Option<crate::parser::types::Location> as Clone>::clone(loc)
+                                    .map(|l| l.line_col),
+                            );
+                            return None;
+                        }
+                    }
+                } // else function has no arguments
+
+                let fn_name = match *fn_target.clone() {
+                    Expr::VarId { name, .. } => name,
+                    _ => {
+                        self.err.type_check_error(
+                            false,
+                            "Function target must be a variable".to_owned(),
+                            &<Option<crate::parser::types::Location> as Clone>::clone(loc)
+                                .map(|l| l.line_col),
+                        );
+                        return None;
+                    }
+                };
+
+                if let Some(id) = self.table.lookup(&fn_name) {
+                    if let Some(Record::Fn { .. }) = self.table.get_record(id) {
+                        // TODO how to get the real Datatype of the Param
+                        todo!();
+                    }
+                }
+
+                None
+            }
+            Expr::Ternary {
+                cond,
+                conseq,
+                alt,
+                loc,
+            } => {
+                let cond_ty = self.visit_expr(cond);
+                if let Some(ty) = cond_ty {
+                    if ty != DataType::Boolean {
+                        self.err.type_check_error(
+                            false,
+                            "Condition must be of type boolean".to_owned(),
+                            &<Option<crate::parser::types::Location> as Clone>::clone(loc)
+                                .map(|l| l.line_col),
+                        );
+                    }
+                }
+                // TODO: should conseq and alt have the same type?
```

Comment:
yep! and the returned type for a ternary should be the type of the conseq and alt